src/libsync/propagatecommonzsync.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) by Ahmed Ammar <ahmed.a.ammar@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#pragma once
+
+#include <QLoggingCategory>
+#include <QTemporaryFile>
+#include <QRunnable>
+#include <QThreadPool>
+
+extern "C" {
+    #include "librcksum/rcksum.h"
+    #include "libzsync/zmap.h"
+    #include "libzsync/sha1.h"
+    #include "libzsync/zsync.h"
+}
+
+namespace OCC {
+Q_DECLARE_LOGGING_CATEGORY(lcZsyncPut)
+Q_DECLARE_LOGGING_CATEGORY(lcZsyncGet)
+
+enum class ZsyncMode { download, upload };
+
+template<typename T>
+    using zsync_unique_ptr = std::unique_ptr<T,std::function<void(T*)>>;
+
+/**
+ * @ingroup libsync
+ *
+ * Helper function to know if we are allowed to attempt using zsync from configuration/command-line options.
+ *
+ */
+bool isZsyncPropagationEnabled(OwncloudPropagator *propagator, const SyncFileItemPtr& item);
+
+/**
+ * @ingroup libsync
+ *
+ * Helper function to get zsync metadata Url.
+ *
+ */
+QUrl zsyncMetadataUrl(OwncloudPropagator *propagator, const QString &path);
+
+/**
+ * @ingroup libsync
+ *
+ * Helper runnable to 'seed' the zsync_state by providing the downloaded metadata and seed file.
+ * This is needed for both upload and download since they both must seed the zsync_state to know which
+ * ranges to upload/download.
+ *
+ */
+class ZsyncSeedRunnable : public QObject, public QRunnable
+{
+    Q_OBJECT
+    QByteArray _zsyncData;
+    QString _zsyncFilePath;
+    QString _tmpFilePath;
+    ZsyncMode _type;
+public:
+    explicit ZsyncSeedRunnable(QByteArray &zsyncData, QString path, ZsyncMode type, QString tmpFilePath = nullptr)
+        : _zsyncData(zsyncData)
+        , _zsyncFilePath(path)
+        , _tmpFilePath(tmpFilePath)
+        , _type(type)
+    {
+    };
+
+    void run();
+
+signals:
+    void finishedSignal(void *zs);
+    void failedSignal(const QString& errorString);
+};
+
+/**
+ * @ingroup libsync
+ *
+ * Helper runnable to generate zsync metadata file when uploading.
+ * Takes an input file path and returns a zsync metadata file path finsihed.
+ *
+ */
+class ZsyncGenerateRunnable : public QObject, public QRunnable
+{
+    Q_OBJECT
+    size_t _blocksize;
```

Comment:
use the inline initialization:

```
size_t _blocksize = 1048576;
off_t _len = 0;
```

So you don't need to repeat that in the constructor.